From bffb97fe859ddde86be2d99cfc77b595fb416498 Mon Sep 17 00:00:00 2001
From: Ray Strode <rstrode@redhat.com>
Date: Sat, 28 Mar 2015 01:05:38 -0400
Subject: seat: drop set_splash function

It's unused.

diff --git a/src/libply-splash-core/ply-seat.c b/src/libply-splash-core/ply-seat.c
index 700332f..6475170 100644
--- a/src/libply-splash-core/ply-seat.c
+++ b/src/libply-splash-core/ply-seat.c
@@ -283,22 +283,6 @@ ply_seat_close (ply_seat_t *seat)
         seat->renderer = NULL;
 }
 
-void
-ply_seat_set_splash (ply_seat_t        *seat,
-                     ply_boot_splash_t *splash)
-{
-        if (seat->splash == splash)
-                return;
-
-        if (seat->splash != NULL)
-                ply_boot_splash_detach_from_seat (splash, seat);
-
-        if (splash != NULL)
-                ply_boot_splash_attach_to_seat (splash, seat);
-
-        seat->splash = splash;
-}
-
 static void
 free_pixel_displays (ply_seat_t *seat)
 {
diff --git a/src/libply-splash-core/ply-seat.h b/src/libply-splash-core/ply-seat.h
index f9ed15d..f84355b 100644
--- a/src/libply-splash-core/ply-seat.h
+++ b/src/libply-splash-core/ply-seat.h
@@ -27,7 +27,6 @@
 #include <stdint.h>
 #include <unistd.h>
 
-#include "ply-boot-splash.h"
 #include "ply-buffer.h"
 #include "ply-event-loop.h"
 #include "ply-keyboard.h"
@@ -53,8 +52,6 @@ void ply_seat_deactivate_renderer (ply_seat_t *seat);
 void ply_seat_activate_renderer (ply_seat_t *seat);
 void ply_seat_refresh_displays (ply_seat_t *seat);
 void ply_seat_close (ply_seat_t *seat);
-void ply_seat_set_splash (ply_seat_t        *seat,
-                          ply_boot_splash_t *splash);
 
 ply_list_t *ply_seat_get_pixel_displays (ply_seat_t *seat);
 ply_list_t *ply_seat_get_text_displays (ply_seat_t *seat);
-- 
cgit v0.10.2

