From 86e8fa74f12e452e1a42e998908873b96b96ebc3 Mon Sep 17 00:00:00 2001
From: Frederic Crozat <fcrozat@suse.com>
Date: Thu, 12 Jun 2014 17:18:59 +0200
Subject: [PATCH 103/119] main: fix typo in debug message

---
 src/main.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/src/main.c b/src/main.c
index ab9e1f9..bc15d65 100644
--- a/src/main.c
+++ b/src/main.c
@@ -836,7 +836,7 @@ plymouth_should_show_default_splash (state_t *state)
         }
 
         if (command_line_has_argument (state->kernel_command_line, "splash=silent")) {
-                ply_trace ("using default splash because kernel command line has option \"splash=slient\"");
+                ply_trace ("using default splash because kernel command line has option \"splash=silent\"");
                 return true;
         }
 
-- 
1.9.0

