diff -Naur plymouth-0.9.3-20160608/scripts/plymouth-update-initrd plymouth-0.9.3-20160608.tpg/scripts/plymouth-update-initrd
--- plymouth-0.9.3-20160608/scripts/plymouth-update-initrd	2016-06-07 12:28:03.000000000 +0000
+++ plymouth-0.9.3-20160608.tpg/scripts/plymouth-update-initrd	2016-06-08 18:30:35.295574949 +0000
@@ -1,2 +1,9 @@
 #!/bin/bash
-dracut -f
+kver=$(uname -r)
+
+if [ -d /lib/modules/${kver} -a -x /usr/bin/kernel-install ]; then
+    /usr/bin/kernel-install remove ${kver} ||:
+    /usr/bin/kernel-install add ${kver} vmlinuz-${kver} ||:
+else
+    dracut -f
+fi
