diff -Naur plymouth-0.9.6-20211010/src/main.c plymouth-0.9.6-20211010.tpg/src/main.c
--- plymouth-0.9.6-20211010/src/main.c	2021-09-27 10:27:11.000000000 +0000
+++ plymouth-0.9.6-20211010.tpg/src/main.c	2021-10-10 16:43:49.162886757 +0000
@@ -2127,7 +2127,7 @@
         state_t state = { 0 };
         int exit_code;
         bool should_help = false;
-        bool no_boot_log = false;
+        bool no_boot_log = true;
         bool no_daemon = false;
         bool debug = false;
         bool ignore_serial_consoles = false;
