diff -Naur plymouth-0.9.4-20200326/configure.ac plymouth-0.9.4-20200326.tpg/configure.ac
--- plymouth-0.9.4-20200326/configure.ac	2020-03-25 19:27:11.000000000 +0000
+++ plymouth-0.9.4-20200326.tpg/configure.ac	2020-03-26 10:33:53.482010298 +0000
@@ -336,6 +336,7 @@
 	   scripts/plymouth-populate-initrd
 	   scripts/plymouth-set-default-theme
            scripts/Makefile
+           systemd-units/plymouth-deactivate.service
            systemd-units/plymouth-halt.service
            systemd-units/plymouth-kexec.service
            systemd-units/plymouth-poweroff.service
diff -Naur plymouth-0.9.4-20200326/systemd-units/Makefile.am plymouth-0.9.4-20200326.tpg/systemd-units/Makefile.am
--- plymouth-0.9.4-20200326/systemd-units/Makefile.am	2020-03-25 19:27:11.000000000 +0000
+++ plymouth-0.9.4-20200326.tpg/systemd-units/Makefile.am	2020-03-26 10:44:59.284014137 +0000
@@ -1,4 +1,5 @@
 systemd_unit_templates =                                                     \
+        plymouth-deactivate.service.in                                       \
         plymouth-switch-root.service.in                                      \
         plymouth-start.service.in                                            \
         plymouth-read-write.service.in                                       \
diff -Naur plymouth-0.9.4-20200326/systemd-units/plymouth-deactivate.service.in plymouth-0.9.4-20200326.tpg/systemd-units/plymouth-deactivate.service.in
--- plymouth-0.9.4-20200326/systemd-units/plymouth-deactivate.service.in	1970-01-01 00:00:00.000000000 +0000
+++ plymouth-0.9.4-20200326.tpg/systemd-units/plymouth-deactivate.service.in	2020-03-26 10:31:03.117009315 +0000
@@ -0,0 +1,10 @@
+[Unit]
+Description=Deactivate Plymouth Boot Screen
+After=rc-local.service plymouth-start.service systemd-user-sessions.service
+Before=getty@tty1.service
+
+[Service]
+ExecStart=-@PLYMOUTH_CLIENT_DIR@/plymouth deactivate
+Type=oneshot
+RemainAfterExit=yes
+TimeoutSec=20
diff -Naur plymouth-0.9.4-20200326/systemd-units/plymouth-quit.service.in plymouth-0.9.4-20200326.tpg/systemd-units/plymouth-quit.service.in
--- plymouth-0.9.4-20200326/systemd-units/plymouth-quit.service.in	2020-03-25 19:27:11.000000000 +0000
+++ plymouth-0.9.4-20200326.tpg/systemd-units/plymouth-quit.service.in	2020-03-26 10:31:50.946009591 +0000
@@ -1,9 +1,10 @@
 [Unit]
 Description=Terminate Plymouth Boot Screen
 After=rc-local.service plymouth-start.service systemd-user-sessions.service
+Before=getty@tty1.service
 
 [Service]
-ExecStart=-@PLYMOUTH_CLIENT_DIR@/plymouth quit
+ExecStart=-@PLYMOUTH_CLIENT_DIR@/plymouth quit --retain-splash
 Type=oneshot
 RemainAfterExit=yes
 TimeoutSec=20
