diff --git a/scripts/plymouth-update-initrd b/scripts/plymouth-update-initrd
index 4ed5709..8ef6bc4 100755
--- a/scripts/plymouth-update-initrd
+++ b/scripts/plymouth-update-initrd
@@ -1,2 +1,8 @@
 #!/bin/bash
-dracut -f
+
+if [ -x "$(command -v kernel-install)" ] && [ -d "/lib/modules/$(uname -r)" ] && [ -x "$(command -v bootctl)" ] && [ "$(bootctl is-installed)" = 'yes' ]; then
+    kernel-install remove $(uname -r) ||:
+    kernel-install add $(uname -r) /boot/vmlinuz-$(uname -r) ||:
+else
+    dracut -f
+fi
