diff -Naur plymouth/src/plymouthd.conf plymouth.tpg/src/plymouthd.conf
--- plymouth/src/plymouthd.conf	2016-05-16 17:21:11.739947468 +0000
+++ plymouth.tpg/src/plymouthd.conf	2016-05-16 19:31:41.677696087 +0000
@@ -1,3 +1,5 @@
 # Administrator customizations go in this file
-#[Daemon]
-#Theme=fade-in
+[Daemon]
+Theme=OpenMandriva
+ShowDelay=0
+DeviceTimeout=5
diff -Naur plymouth/src/plymouthd.defaults plymouth.tpg/src/plymouthd.defaults
--- plymouth/src/plymouthd.defaults	2016-05-16 17:21:11.739947468 +0000
+++ plymouth.tpg/src/plymouthd.defaults	2016-05-16 19:31:20.849126400 +0000
@@ -2,5 +2,5 @@
 # upgrades.
 [Daemon]
 Theme=spinner
-ShowDelay=5
+ShowDelay=0
 DeviceTimeout=5
