diff -Naur plymouth-0.9.0/src/libply/ply-logger.c plymouth-0.9.0.tpg/src/libply/ply-logger.c
--- plymouth-0.9.0/src/libply/ply-logger.c	2014-11-08 12:53:09.000000000 +0000
+++ plymouth-0.9.0.tpg/src/libply/ply-logger.c	2014-11-08 12:55:01.722168938 +0000
@@ -579,6 +579,16 @@
 
         return logger->tracing_is_enabled != false;
 }
+#else
+void
+ply_logger_toggle_tracing (ply_logger_t *logger)
+{
+}
+bool
+ply_logger_is_tracing_enabled (ply_logger_t *logger)
+{
+    return false;
+}
 #endif /* PLY_ENABLE_TRACING */
 
 /* vim: set ts=4 sw=4 expandtab autoindent cindent cino={.5s,(0: */
diff -Naur plymouth-0.9.0/src/libply/ply-logger.h plymouth-0.9.0.tpg/src/libply/ply-logger.h
--- plymouth-0.9.0/src/libply/ply-logger.h	2014-11-08 12:53:09.000000000 +0000
+++ plymouth-0.9.0.tpg/src/libply/ply-logger.h	2014-11-08 12:56:48.765118570 +0000
@@ -77,11 +77,15 @@
 
 /* tracing is a debugging facility that incurs a hefty performance hit on the
  * program, so we conditionally compile support for it
+ * We unconditionally need to define the symbols though (can't just define them
+ * to nop and false), because code linking to the library shouldn't have ABIs
+ * pulled out from under it...
  */
-#ifdef PLY_ENABLE_TRACING
+
 void ply_logger_toggle_tracing (ply_logger_t *logger);
 bool ply_logger_is_tracing_enabled (ply_logger_t *logger);
 
+#ifdef PLY_ENABLE_TRACING
 #define ply_logger_trace(logger, format, args ...)                              \
         do                                                                             \
         {                                                                            \
@@ -101,8 +105,6 @@
         while (0)
 #else
 #define ply_logger_trace(logger, format, args ...)
-#define ply_logger_toggle_tracing(logger)
-#define ply_logger_is_tracing_enabled(logger) (false)
 #endif /* PLY_ENABLE_TRACING */
 
 /* convenience macros
