--- a/scripts/generate-version.sh
+++ b/scripts/generate-version.sh
@@ -1,7 +1,6 @@
 #!/bin/sh
 exec 3>&2 2> /dev/null
 SRCDIR=$(dirname "$0")/..
-cd "$SRCDIR"
 CWD=$(realpath "$PWD")
 exec 2>&3
 
@@ -12,7 +11,7 @@
 
 # If it's not from a git checkout, assume it's from a tarball
 if ! git rev-parse --is-inside-git-dir > /dev/null 2>&1; then
-    VERSION_FROM_DIR_NAME=$(basename "$CWD" | sed -n 's/^plymouth-\([^-]*\)$/\1/p')
+    VERSION_FROM_DIR_NAME=$(basename "$CWD" | sed -n 's/^plymouth-\([^-].*\)$/\1/p')
 
     if [ -n "$VERSION_FROM_DIR_NAME" ]; then
         echo "$VERSION_FROM_DIR_NAME"
