diff -Naur a/configure.ac b/configure.ac
--- a/configure.ac	2021-04-06 11:53:40.000000000 +0000
+++ b/configure.ac	2021-04-14 10:05:09.765211341 +0000
@@ -333,6 +333,7 @@
 	   scripts/plymouth-populate-initrd
 	   scripts/plymouth-set-default-theme
            scripts/Makefile
+           systemd-units/plymouth-deactivate.service
            systemd-units/plymouth-halt.service
            systemd-units/plymouth-kexec.service
            systemd-units/plymouth-poweroff.service
diff -Naur a/systemd-units/Makefile.am b/systemd-units/Makefile.am
--- a/systemd-units/Makefile.am	2021-04-06 11:53:40.000000000 +0000
+++ b/systemd-units/Makefile.am	2021-04-14 10:05:38.513226371 +0000
@@ -1,4 +1,5 @@
 systemd_unit_templates =                                                     \
+        plymouth-deactivate.service.in                                       \
         plymouth-switch-root.service.in                                      \
         plymouth-switch-root-initramfs.service.in                            \
         plymouth-start.service.in                                            \
diff -Naur a/systemd-units/plymouth-deactivate.service.in b/systemd-units/plymouth-deactivate.service.in
--- a/systemd-units/plymouth-deactivate.service.in	1970-01-01 00:00:00.000000000 +0000
+++ b/systemd-units/plymouth-deactivate.service.in	2021-04-14 10:06:08.494255512 +0000
@@ -0,0 +1,10 @@
+[Unit]
+Description=Deactivate Plymouth Boot Screen
+After=rc-local.service plymouth-start.service systemd-user-sessions.service
+Before=getty@tty1.service
+
+[Service]
+ExecStart=-@PLYMOUTH_CLIENT_DIR@/plymouth deactivate
+Type=oneshot
+RemainAfterExit=yes
+TimeoutSec=20
diff -Naur a/systemd-units/plymouth-quit.service.in b/systemd-units/plymouth-quit.service.in
--- a/systemd-units/plymouth-quit.service.in	2021-04-06 11:53:40.000000000 +0000
+++ b/systemd-units/plymouth-quit.service.in	2021-04-14 10:11:26.025103567 +0000
@@ -1,9 +1,10 @@
 [Unit]
 Description=Terminate Plymouth Boot Screen
 After=rc-local.service plymouth-start.service systemd-user-sessions.service
+Before=getty@tty1.service
 
 [Service]
-ExecStart=-@PLYMOUTH_CLIENT_DIR@/plymouth quit
+ExecStart=-@PLYMOUTH_CLIENT_DIR@/plymouth quit --retain-splash
 Type=oneshot
 RemainAfterExit=yes
 TimeoutSec=20
