diff -Naur a/scripts/plymouth-update-initrd b/scripts/plymouth-update-initrd
--- a/scripts/plymouth-update-initrd	2021-04-06 11:53:40.000000000 +0000
+++ b/scripts/plymouth-update-initrd	2021-04-14 20:45:32.600009773 +0000
@@ -1,2 +1,8 @@
 #!/bin/bash
-dracut -f
+
+if [ -x "$(command -v kernel-install)" ] && [ -d "/lib/modules/$(uname -r)" ] && [ -x "$(command -v bootctl)" ] && [ "$(bootctl is-installed)" = 'yes' ]; then
+    kernel-install remove $(uname -r) ||:
+    kernel-install add $(uname -r) /boot/vmlinuz-$(uname -r) ||:
+else
+    dracut -f
+fi
